### PR TITLE
Add redirect paths for authentication

### LINKED ISSUE #283: Redirect authentication for microsoft in same page and don't through pop-up.
When click on button for authentication on Microsoft or when access the "view" with tag @login_required don't open the pop-up, but redirect in same page.

This solution is more clear and can be used in many ways.

### PR DESCRIPTION
**Pull request by kangasta**
Add `to-auth-redirect` and `from-auth-redirect` paths to url configuration to allow authentication without opening new window.

Fixes #283

### CODE REVIEW

**Comment by kangasta** [comment]
Added unittests for the redirect views.

The Travis build seems to fail because of Github request rate limits:
```
Cloning into 'AngellusMortis/django_microsoft_auth'...

fatal: unable to access 'https://github.com/AngellusMortis/django_microsoft_auth.git/': The requested URL returned error: 429
```

Those failing jobs could be re-triggered manually, but I don't have rigths to do so.

**Comment by sagarc03** [comment]
@kangasta can you rebase with master try to push again?

**Comment by AngellusMortis** [comment]
Personally I would love two see these two features split up. I am 100% okay with adding an additional view to allow you to do a single page flow for auth instead of a popup flow.

However, I still am not sure on removing the `sites` dependency. The `sites` framework comes with Django. It is the recommended way in the Django docs to handle absolute URLs. If you want to win me over on this feature, I need some good justification for it.

**Comment by kangasta** [comment]
I agree that this would have been cleaner as two separate pull requests.

In my setup, the `sites` framework was only used to provide the public URL for this app, which in my opinion added extra complexity to the deployment process.

Optional `sites` dependency would make configuring the public URL simpler in situations where the Django instance is serving only one site. `MICROSOFT_AUTH_DOMAIN` parameter would allow providing the public URL in, for example, an environment variable instead of having to install `sites` framework and modify a DB row. The sites framework could still be used, the goal here would be to add an alternative for users without other need for the `sites` framework.

**Comment by AngellusMortis** [comment]
> which in my opinion added extra complexity to the deployment process.

I definitely agree the sites framework can add some complexity, but for better or worse, it is the way handling public absolute URLs works in Django. I would rather not have two ways to solve the same problem. The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) says "There should be one-- and preferably only one --obvious way to do it."

If there was a reason the sites framework just does not work for you, I would be willing to reconsider.

Also, something else you can do to help the complexity issue for deployments is you can create a data migration to fix automatically fix the default site name. That would also allow you to still version control the site name. It is actually what the [Django docs](https://docs.djangoproject.com/en/3.1/ref/contrib/sites/#enabling-the-sites-framework) recommend doing:

> To set the correct name and domain for your project, you can use a data migration.

**Comment by kangasta** [comment]
Yes, I see the risk of confusion here if there are two possible options for setting the public domain.

I will edit this pull request to only contain the changes to add redirect views.

File: microsoft_auth/context_processors.py [hunk 1]
```diff
@@ -41,10 +41,15 @@ def microsoft(request):
...
-    state = signer.sign(get_token(request))
-    microsoft = MicrosoftClient(state=state, request=request)
+    # Initialize Microsoft client using dict with CSRF token and optional
+    # next path as state variable
+    next_ = request.GET.get("next")
+    state = dict(token=get_token(request))
+    if next_:
+        state["next"] = next_
+
+    signed_state = dumps(state, salt="microsoft_auth")
```
Comment (AngellusMortis) [inline microsoft_auth/context_processors.py:51]: Very nice. I saw the email notification for this and I was like "oh no, here we go again", but signing the redirect is a great way to handle this.

**Comment by alfonsrv** [comment]
What's the point exactly of having both an `from-auth-redirect` and `auth-callback` and url? To me the latter seems redundant, since the request has already been verified in the initial `from-auth-redirect` call – might as well handle the desired logic there. Also removes the requirement of having to parse the requests twice.

**Pull request merged**